core/src/test/java/org/wso2/msf4j/InterceptorTestBase.java

```diff
@@ -1,61 +1,167 @@
-            headerMap.put(keyValues[i], keyValues[i + 1]);
+    /**
+     * @param path      http path
+     * @param rawData   data to post
+     * @param keepAlive should the connection be kept alive
+     * @param tClass    type of the expected object
+     * @param headers   request headers
+     * @param <T>       type of the expected object
+     * @return object from response
+     * @throws IOException on any exception
+     */
+    protected <T> T doPostAndGetResponseObject(String path, String rawData, boolean keepAlive, Class<T> tClass,
+                                               Map<String, String> headers) throws IOException {
+        HttpURLConnection urlConn = request(path, HttpMethod.POST, keepAlive, headers);
+        ByteBuffer encodedData = Charset.defaultCharset().encode(rawData);
+        urlConn.setRequestMethod("POST");
+        urlConn.setRequestProperty("Content-Type", MediaType.APPLICATION_FORM_URLENCODED);
+        urlConn.setRequestProperty("Content-Length", String.valueOf(encodedData.array().length));
+        try (OutputStream os = urlConn.getOutputStream()) {
+            os.write(Arrays.copyOf(encodedData.array(), encodedData.limit()));
         }
-        return doGet(resource, headerMap);
+        return getResponseObject(urlConn, tClass);
     }
 
-    private int doGet(String resource, Map<String, String> headers) throws Exception {
-        URL url = baseURI.resolve(resource).toURL();
+    /**
+     * Get java object from http url connection.
+     *
+     * @param urlConn http url connection
+     * @param tClass  type of the expected object
+     * @param <T>     type of the expected object
+     * @return object from response
+     * @throws IOException on any exception
+     */
+    protected <T> T getResponseObject(HttpURLConnection urlConn, Class<T> tClass) throws IOException {
+        return gson.fromJson(getResponseString(urlConn), tClass);
+    }
+
+    /**
+     * Get java object from http url connection.
+     *
+     * @param urlConn http url connection
+     * @return string from response
+     * @throws IOException on any exception
+     */
+    protected String getResponseString(HttpURLConnection urlConn) throws IOException {
+        InputStream inputStream = urlConn.getInputStream();
+        String response = StreamUtil.asString(inputStream);
+        IOUtils.closeQuietly(inputStream);
+        urlConn.disconnect();
+        return response;
+    }
+
+    /**
+     * Get http url connection for the path and method specified.
+     *
+     * @param path      http path
+     * @param method    http method
+     * @param keepAlive should the connection be kept alive
+     * @param headers   request headers
+     * @return http url connection instance
+     * @throws IOException on error creating http url connection
+     */
+    private HttpURLConnection request(String path, String method, boolean keepAlive, Map<String, String> headers)
```

Comment:
IMO the method name "request" is not self-explanatory. Someone might need to read the method description to understand its purpose. Shall we rename this to something similar to createHttpURLConnection()?